### PR TITLE
[No QA] Add support for Pusher responses

### PR DESCRIPTION
**Pull request by Gonals**
### Details
We haven't migrated any commands yet, so we need to add some changes in Web to test this.

### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/211236

### Tests
1. In web-E, you'll need to add an Onyx response to a command to test this. I simply added this line [here](https://github.com/Expensify/Web-Expensify/blob/80574708de8162aa09c17e6fa297c31e787d068c/api.php#L607), so it would be triggered when a new comment is added on the chat:
```
$onyxUpdate[] = ['onyxMethod' => 'Report_AddCommentSet', 'key' =>'Nepal', 'value' => 'Test'];
```
2. You can just log the JSON to the console [here](https://github.com/Expensify/App/pull/9219/files#diff-93c1bb3a403f942b2809d43de01f06366a9fed3060e3894959c268b53eb97be8R309) to make sure it is received.
3. After that, open a chat in NewDot and send a message. In a second or two you should see the response showing up in the console. 
4. Tail the logs.
5. Remove or change one of the keys in onyxUpdate and send a message to the chat. The response that shows up in the console should now be an empty array, and a warning indicating that the response was not correctly formatted should show in the logs.

### CODE REVIEW

File: src/libs/Network/SequentialQueue.js [hunk 1]
```diff
@@ -98,8 +104,19 @@ function push(request) {
     flush();
 }
 
+/**
+ * @returns {Promise}
+ */
+function getCurrentRequest() {
+    if (currentRequest === null) {
+        return Promise.resolve();
+    }
```
Comment (marcaaron) [inline src/libs/Network/SequentialQueue.js:113]: NAB, could also initialize/reset with a `Promise.resolve()` but this works.

File: src/libs/Pusher/EventType.js [hunk 2]
```diff
@@ -11,4 +11,6 @@ export default {
     PREFERRED_LOCALE: 'preferredLocale',
     EXPENSIFY_CARD_UPDATE: 'expensifyCardUpdate',
     SCREEN_SHARE_REQUEST: 'screenshareRequest',
+    ONYX_API_UPDATE: 'onyxApiUpdate',
+    ONYX_API_UPDATE_CHUNK: 'chunked-onyxApiUpdate',
```
Comment (marcaaron) [inline src/libs/Pusher/EventType.js:15]: After [these changes](https://github.com/Expensify/App/pull/9203/files#r883952420
) we won't need the prepend chunked to channel names. 

cc @danieldoglas to confirm

File: src/libs/Pusher/pusher.js [hunk 3]
```diff
@@ -358,6 +360,46 @@ function reconnect() {
...
 
+/**
+ * Abstraction around subscribing to private user channel events. Handles all logs and errors automatically.
+ *
+ * @param {String} eventName
+ * @param {String} accountID
+ * @param {Function} onEvent
+ * @param {Boolean} isChunked
+ */
+function subscribeToPrivateUserChannelEvent(eventName, accountID, onEvent, isChunked = false) {
```
Comment (marcaaron) [inline src/libs/Pusher/pusher.js:371]: The isChunked possible conflict with the PR mentioned before

File: src/libs/Pusher/pusher.js [hunk 4]
```diff
@@ -358,6 +360,46 @@ function reconnect() {
...
+
+    /**
+     * @param {Object} pushJSON
+     */
+    function logPusherEvent(pushJSON) {
+        Log.info(`[Report] Handled ${eventName} event sent by Pusher`, false, pushJSON);
+    }
+
+    function onPusherResubscribeToPrivateUserChannel() {
+        NetworkConnection.triggerReconnectionCallbacks('Pusher re-subscribed to private user channel');
```
Comment (marcaaron) [inline src/libs/Pusher/pusher.js:382]: **opinion (non-blocking):** It stuck out to me that the Pusher library is now triggering and aware of the reconnection callbacks. Feels like it now has extra responsibility it should not have. I would maybe put this into something like a `PusherUtils` or just have it in `User` eventually (after the usages in `Report.js` are cleaned up). But not sure we need to do this now.

**Review by marcaaron — DISMISSED** [review]

File: src/libs/Pusher/EventType.js (hunk 2 above)
Comment (Gonals) [reply src/libs/Pusher/EventType.js:15]: Oh nice! That one seems fully approved, so it should be merging today, right @alex-mechler?

File: src/libs/Pusher/pusher.js [hunk 5]
```diff
@@ -358,6 +360,46 @@ function reconnect() {
...
+/**
+ * Abstraction around subscribing to private user channel events. Handles all logs and errors automatically.
+ *
+ * @param {String} eventName
+ * @param {String} accountID
+ * @param {Function} onEvent
+ * @param {Boolean} isChunked
+ */
+function subscribeToPrivateUserChannelEvent(eventName, accountID, onEvent, isChunked = false) {
+    const pusherChannelName = `private-encrypted-user-accountID-${accountID}${CONFIG.PUSHER.SUFFIX}`;
```
Comment (yuwenmemon) [inline src/libs/Pusher/pusher.js:372]: NAB: We should consider making this prefix a constant like we do in Web: https://github.com/Expensify/Web-Expensify/blob/25038455beaab04f40fe01f595608655d67e5a96/site/lib/PushNotifications.jsx#L36-L42

File: src/libs/Pusher/pusher.js (hunk 4 above)
Comment (yuwenmemon) [reply src/libs/Pusher/pusher.js:382]: I like putting it into `PusherUtils` :+1:

**Review by yuwenmemon — DISMISSED** [review]

File: src/libs/Pusher/pusher.js (hunk 4 above)
Comment (Gonals) [reply src/libs/Pusher/pusher.js]: Moved!

File: src/libs/Pusher/EventType.js (hunk 2 above)
Comment (alex-mechler) [reply src/libs/Pusher/EventType.js:15]: Yup, that sounds fine. I have to remove the other `chunked-` channels later from this repo anyways.

**Review by yuwenmemon — APPROVED** [review]

**Comment by yuwenmemon** [comment]
@marcaaron all yours

**Review by marcaaron — APPROVED** [review]

**Comment by marcaaron** [comment]
🚀

File: src/libs/actions/Report.js [hunk 6]
```diff
@@ -723,45 +723,6 @@ function getReportChannelName(reportID) {
...
-    }
-
-    /**
-     * @param {*} error
-     */
-    function onSubscriptionFailed(error) {
-        Log.hmmm('[Report] Failed to subscribe to Pusher channel', false, {error, pusherChannelName, eventName});
-    }
-
-    Pusher.subscribe(pusherChannelName, eventName, onEventPush, onPusherResubscribeToPrivateUserChannel)
```
Comment (mountiny) [inline src/libs/actions/Report.js:761]: Here there was not the `isChunked` but in the reportUtils it is back.

File: src/libs/actions/User.js [hunk 7]
```diff
@@ -302,6 +304,18 @@ function subscribeToUserEvents() {
 
     const pusherChannelName = `private-encrypted-user-accountID-${currentUserAccountID}${CONFIG.PUSHER.SUFFIX}`;
 
+    // Receive any relevant Onyx updates from the server
+    PusherUtils.subscribeToPrivateUserChannelEvent(Pusher.TYPE.ONYX_API_UPDATE, currentUserAccountID, (pushJSON) => {
+        SequentialQueue.getCurrentRequest().then(() => {
```
Comment (roryabraham) [inline src/libs/actions/User.js:309]: Hey everyone 👋🏼 I know it's been a little while, but can someone explain why we needed to add `SequentialQueue.getCurrentRequest()` here? It's not clear to me what it's doing.

Comment (marcaaron) [reply src/libs/actions/User.js:309]: The first I heard about it you were the one proposing it 😄  

https://expensify.slack.com/archives/C02HWMSMZEC/p1652731618392499?thread_ts=1652466984.820589&cid=C02HWMSMZEC

Comment (roryabraham) [reply src/libs/actions/User.js:309]: I think I understand why this was added now, I tried to summarize it as concisely as possible [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1681147855907809?thread_ts=1679595548.409089&cid=C01GTK53T8Q)

Comment (marcaaron) [reply src/libs/actions/User.js:309]: It was always a longer goal of mine to add a `NETWORK.md` and that explanation would be perfect for it (or added as a comment here).

**Pull request merged**